velox/substrait/SubstraitExtensionCollector.cpp

```diff
@@ -0,0 +1,137 @@
+      uris[functionId.uri] = extensionUri;
+    } else {
+      extensionUri = uri->second;
+    }
+
+    auto extensionFunction =
+        plan->add_extensions()->mutable_extension_function();
+    extensionFunction->set_extension_uri_reference(
+        extensionUri->extension_uri_anchor());
+    extensionFunction->set_function_anchor(referenceNum);
+    extensionFunction->set_name(functionId.signature);
+  }
+}
+
+void SubstraitExtensionCollector::addExtensionTypesToPlan(
+    ::substrait::Plan* plan) const {
+  using SimpleExtensionURI = ::substrait::extensions::SimpleExtensionURI;
+  int uriPos = 1;
+  std::unordered_map<std::string, SimpleExtensionURI*> uris;
+  for (auto& [referenceNum, typeId] : extensionTypes_->forwardMap_) {
+    SimpleExtensionURI* extensionUri;
+    const auto uri = uris.find(typeId.uri);
+    if (uri == uris.end()) {
+      extensionUri = plan->add_extension_uris();
+      extensionUri->set_extension_uri_anchor(++uriPos);
+      extensionUri->set_uri(typeId.uri);
+      uris[typeId.uri] = extensionUri;
+    } else {
+      extensionUri = uri->second;
+    }
+
+    auto extensionType = plan->add_extensions()->mutable_extension_type();
+    extensionType->set_extension_uri_reference(
+        extensionUri->extension_uri_anchor());
+    extensionType->set_type_anchor(referenceNum);
+    extensionType->set_name(typeId.name);
+  }
+}
+
+std::optional<int> SubstraitExtensionCollector::getFunctionReference(
+    const std::string& functionName,
+    const std::vector<TypePtr>& arguments) {
+  const auto& functionId =
+      extensionIdResolver_->resolve(functionName, arguments);
+  const auto& anchorReference =
+      extensionFunctions_->reverseMap_.find(functionId);
+  if (anchorReference != extensionFunctions_->reverseMap_.end()) {
+    return std::make_optional(anchorReference->second);
+  }
+  ++functionReference_;
+  extensionFunctions_->put(functionReference_, functionId);
+  return std::make_optional(functionReference_);
+}
+
+template <typename T>
+void SubstraitExtensionCollector::BiDirectionHashMap<T>::put(
+    const int& key,
+    const T& value) {
+  forwardMap_[key] = value;
+  reverseMap_[value] = key;
+}
+
+void SubstraitExtensionCollector::addExtensionsToPlan(
+    ::substrait::Plan* plan) const {
+  addExtensionFunctionsToPlan(plan);
+}
+std::optional<int> SubstraitExtensionCollector::getTypeReference(
+    const TypePtr& type) {
+  const auto& typeId = extensionIdResolver_->resolve(type);
+  const auto& typeReference = extensionTypes_->reverseMap_.find(typeId);
+  if (typeReference != extensionTypes_->reverseMap_.end()) {
+    return std::make_optional(typeReference->second);
+  }
+  ++functionReference_;
+  extensionTypes_->put(functionReference_, typeId);
+  return std::make_optional(functionReference_);
+}
+
+FunctionId ExtensionIdResolver::resolve(
+    const std::string& funcName,
+    const std::vector<TypePtr>& arguments) const {
+  if (!arguments.empty()) {
```

Comment:
This code will be easier to read if the check is reversed:

```
if (arguments.empty()) {
   return {"", funcName};
}

...
```

Reply:
Done